README.md

```diff
@@ -0,0 +1,121 @@
+
+## Assumptions
+
+  * kubectl apply will be modified to accurately [report if changes had to be applied](https://github.com/kubernetes/kubernetes/pull/52647).
+
+# Target Ansible
+
+```yaml
+- kubectl:
+
+    # State present implies kubectl apply, and absent implies delete. Many options
+    # below would not be relevant for delete.
+    state: present
+
+    kubeconfig:
+      # Kubeconfig file on the remote host:
+      file: /etc/origin/master/admin.kubeconfig
+      # Alternatively:
+      inline: "{{ lookup('file', 'local_on_control_host.kubeconfig') }}"
+
+    # Optional, omit if your configuration specifies a namespace, or you're operating on
+    # cluster level objects.
+    namespace: default
+
+    # You can specify your kubernetes config inline if desired. The | is important here:
+    inline: |
+      kind: "Namespace"
+      apiVersion: v1
+      metadata:
+        name: testnamespace
+
+    # Alternatively you can use files that live with your role/playbook. This would be
+    # mutually exclusive with inline data.
+    # Each file is copied over to the remote host, potentially edited or patched, and
+    # then submitted with apply.
+    # Specifying multiple files implies one execution of kubectl apply. If you want multiple
+    # you can use separate tasks or a with_items loop.
+    files:
+    - src: configmap.json
+    - src: subdir/
+    - src: service.yml
+      # Support small edits to the yaml on the fly, without going to a full template. Operates
+      # on the temporary copy of the file sent to the remote host.
+      yedit:
+        key: spec.metadata.name
+        value: "{{ replacement_name }}"
+        value_type: list
+    - src: patchme.yml
+      # Support supplying structured patches with "kubectl patch" prior to submitting to the server.
+      patches:
+      - patch1.yml
+      - patch2.yml
+    # This would be nice too, kubectl supports, would likely imply no yediting or patching unless
+    # we want the module to be in the business of fetching files.
+    - src: https://github.com/me/project/template.yml
+
+    # Support templating files out automatically. Use Ansible's native mechanisms for jinja
+    # templates but don't require the user to copy the file and manually clean it up with
+    # their own tasks.
+    templates:
+    - src: apptemplate.yml
+      vars:
+        foo: bar
+
+    # Control what binary to use, as oc is required for apply operations on OpenShift types:
+    binary: oc
+
+    # Various other kubectl options could be supported:
+    prune: true
+    dry_run: true
+    overwrite: true
+    selector: something=true
+
+
+# kubectl delete is exposed as state absent:
+- kubectl:
+    kubeconfig:
+      file: /etc/origin/master/admin.kubeconfig
+    state: absent
+    files:
+    - src: subdir/
+    - src: configmap.json
+    inline: |
+      kind: "Namespace"
+      apiVersion: v1
+      metadata:
+        name: testnamespace
+
+
+- kubectl_facts:
```

Comment:
For this, any reason to bias towards `kubectl` vs `kubernetes_facts`?  The implementation is somewhat arbitrary (in the future would be python)

Reply:
That's interesting. The thought was to clearly identify that this is talking to kubectl on the target host, not talking to the actual Kube REST API from the target host. Could we just flip users from one to the other in the future? It seems like it *might* be theoretically possible but I'm not sure it's the kind of thing we could do without upsetting people, although in essence it's not much different, still the same host talking to the same REST API at the end of the day.

Reply:
I was assuming 'kind' was case insensitive as we do in some of these pre-existing modules or on CLI, are you commenting specifically that it should be case sensitive, or that "kind" literally should not be an option here, i.e. should it be:

```
kind: ConfigMap
resource: myconfig (assuming you want a specific resource not a list)
```